packages/run-protocol/src/vpool-xyk-amm/pool.js

```diff
@@ -1,301 +1,328 @@
+ * fromCentralPriceAuthority: PriceAuthority,
+ * quoteIssuerKit: IssuerKit,
+ * zcf: ZCF,
+ * timer: TimerService,
+ * paramAccessor,
+ * }>} ImmutableState
+ *
+ * @typedef {{
+ * updater: IterationObserver<any>,
+ * notifier: Notifier<any>,
+ * poolSeat: ZCFSeat,
+ * liqTokenSupply: bigint,
+ * }} MutableState
+ *
+ * @typedef {{
+ *   state: ImmutableState & MutableState,
+ *   facets: {
+ *     helper: import('@agoric/vat-data/src/types').KindFacet<typeof helperBehavior>,
+ *     pool: import('@agoric/vat-data/src/types').KindFacet<any>, // FIXME typeof poolBehavior
+ *     singlePool: import('@agoric/vat-data/src/types').KindFacet<VirtualPool>,
+ *   },
+ * }} MethodContext
+ */
+
+const updateUpdaterState = (updater, pool) =>
+  // TODO: when governance can change the interest rate, include it here
+  updater.updateState({
+    central: pool.getCentralAmount(),
+    secondary: pool.getSecondaryAmount(),
+  });
+
+const helperBehavior = {
+  /** @type {import('@agoric/vat-data/src/types').PlusContext<MethodContext, AddLiquidityActual>} */
+  addLiquidityActual: (
+    { state },
+    pool,
+    zcfSeat,
+    secondaryAmount,
+    poolCentralAmount,
+    feeSeat,
+  ) => {
+    const { poolSeat, liquidityBrand, liquidityZcfMint, updater, zcf } = state;
+
+    // addLiquidity can't be called until the pool has been created. We verify
+    // that the asset is NAT before creating a pool.
+
+    const liquidityValueOut = calcLiqValueToMint(
+      state.liqTokenSupply,
+      // @ts-expect-error NatValue cast
+      zcfSeat.getStagedAllocation().Central.value,
+      poolCentralAmount.value,
+    );
+
+    const liquidityAmountOut = AmountMath.make(
+      liquidityBrand,
+      liquidityValueOut,
+    );
+    liquidityZcfMint.mintGains(
+      harden({ Liquidity: liquidityAmountOut }),
+      poolSeat,
+    );
+    state.liqTokenSupply += liquidityValueOut;
+
+    poolSeat.incrementBy(
+      zcfSeat.decrementBy(
+        harden({
+          Central: zcfSeat.getStagedAllocation().Central,
+          Secondary: secondaryAmount,
+        }),
+      ),
+    );
+
+    zcfSeat.incrementBy(
+      poolSeat.decrementBy(harden({ Liquidity: liquidityAmountOut })),
+    );
+    if (feeSeat) {
+      zcf.reallocate(poolSeat, zcfSeat, feeSeat);
+    } else {
+      zcf.reallocate(poolSeat, zcfSeat);
+    }
+    zcfSeat.exit();
+    updateUpdaterState(updater, pool);
+    return 'Added liquidity.';
+  },
+};
+
+const poolBehavior = {
+  getLiquiditySupply: ({ state: { liqTokenSupply } }) => liqTokenSupply,
+  getLiquidityIssuer: ({ state: { liquidityIssuer } }) => liquidityIssuer,
+  getPoolSeat: ({ state: { poolSeat } }) => poolSeat,
+  getCentralAmount: ({ state: { poolSeat, centralBrand } }) =>
+    poolSeat.getAmountAllocated('Central', centralBrand),
+  getSecondaryAmount: ({ state }) =>
+    state.poolSeat.getAmountAllocated('Secondary', state.secondaryBrand),
+
+  addLiquidity: ({ state, facets: { helper, pool } }, zcfSeat) => {
```

Comment:
Not gonna block on this but it would help future refactoring to have types here. I tried it and there's a gotcha with circular references, so that the `facets` can't be destructured. There's also a circular definition of the return value so that needs to be specified,
```suggestion
  /**
   * @param {MethodContext} context
   * @param {ZCFSeat} zcfSeat
   * @returns {string}
   */
  addLiquidity: ({ state, facets }, zcfSeat) => {
    const { helper, pool } = facets;
```

Reply:
done